src/test/java/com/microsoft/sqlserver/jdbc/NTLMConnectionTest.java

```diff
@@ -0,0 +1,689 @@
+                    byte[] badComputer = {0, 0, 0, 0, 0, 0, 0, 0, 0};
+                    byte[] badToken = getChallengeToken(NTLM_CHALLENGE_MSVAVDNSCOMPUTERNAME_OFFSET, badComputer);
+
+                    auth.generateClientContext(badToken, done);
+
+                } catch (Exception e) {
+                    // this should just generate a warning but not fail
+                    fail(e.getMessage());
+                }
+
+                // verify message was logged
+                handler.flush();
+                assertTrue(loggerContent.toString().matches(TestUtils.formatErrorMsg("R_ntlmBadComputer")));
+            }
+        }
+    }
+
+    @Test
+    public void testNTLMBadAvid() throws SQLException {
+        if (ntlmPropsDefined) {
+            try (SQLServerConnection con = (SQLServerConnection) DriverManager.getConnection(connectionStringNTLM)) {
+                getServerFqdn(con);
+                SSPIAuthentication auth = new NTLMAuthentication(con, serverFqdn, "domainName", "userName", "password",
+                        "hostname");
+                boolean[] done = {false};
+                try {
+                    byte[] badAvid = {-1, 0};
+                    byte[] badToken = getChallengeToken(NTLM_CHALLENGE_TARGETINFO_OFFSET, badAvid);
+
+                    auth.generateClientContext(badToken, done);
+
+                } catch (Exception e) {
+                    assertTrue(e.getMessage().matches(TestUtils.formatErrorMsg("R_ntlmUnknownValue")));
+                }
+            }
+        }
+    }
+
+    @Test
+    public void testNTLMBadTimestamp() throws SQLException {
+        if (ntlmPropsDefined) {
+            loggerContent.reset();
+
+            try (SQLServerConnection con = (SQLServerConnection) DriverManager.getConnection(connectionStringNTLM)) {
+                getServerFqdn(con);
+                SSPIAuthentication auth = new NTLMAuthentication(con, serverFqdn, "domainName", "userName", "password",
+                        "hostname");
+                boolean[] done = {false};
+
+                try {
+
+                    byte[] badTimestamp = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0};
+                    byte[] badToken = getChallengeToken(NTLM_CHALLENGE_MSVAVTIMESTAMP_OFFSET, badTimestamp);
+
+                    auth.generateClientContext(badToken, done);
+
+                } catch (Exception e) {
+                    // this should just generate a warning but not fail
+                    fail(e.getMessage());
+                }
+            }
+
+            // verify message was logged
+            handler.flush();
+            assertTrue(loggerContent.toString().matches(TestUtils.formatErrorMsg("R_ntlmNoTimestamp")));
+        }
+    }
+
+    /*
+     * Get Server FQDN from connection
+     */
+    private void getServerFqdn(SQLServerConnection con) throws SQLException {
+        try {
+            serverFqdn = InetAddress.getByName(con.currentConnectPlaceHolder.getServerName()).getCanonicalHostName()

```

Comment:
Why do we have to do this? Can't we extract server name from the connection string and just get the `FQDN` of it?

Reply:
same thing, but ok